Add license header to every Python file

yguclu opened the pull request:
- Create `script/` folder for maintenance scripts
- Create `script/add_header.py` for adding a license header to the top of every Python file
- Apply script to `examples/`, `mesh/`, `performance/`, and `psydac/` folders
- Remove obsolete header lines (e.g. `# coding utf-8`, `#! /usr/bin/python`, `# Copyright ...`) from all Python files
- Clean up imports in several modules across the library

FrederikSchnack commented:
Is it necessary to have this script in the repo? I feel like this is a one-time use and all new files should be created already having the correct header.

EmilyBourne commented:
> Is it necessary to have this script in the repo? I feel like this is a one-time use and all new files should be created already having the correct header.

Should being the operative word unless you enforce its use. You might want to try: https://github.com/fsfe/reuse-action

yguclu commented:
> With these changes, could we make the headers of all files the same as well? Some of them have a `coding utf-8` in many different formats, others some `#! /usr/bin/python` magic commands and others have a custom copyright on top.
> 
> Is there any official convention on how file headers should look like?

1. The `#coding utf-8` line is a remnant from Python 2, hence it should be removed. Python 3 defaults to UTF-8 as far as I know.
2. The `#! /usr/bin/python` line is only useful if a script is executable, and I think it should also be removed.
3. Custom copyright info should also be removed, and superseded by the new headers.

yguclu commented:
> Is it necessary to have this script in the repo? I feel like this is a one-time use and all new files should be created already having the correct header.

Not necessary, but in my experience it is useful to keep any maintenance script "just in case". I expect that we will need other scripts in the future, including submission scripts for the supercomputers. Hence I believe it is useful to have a folder to contain them all. Please note that the `script` folder should not be installed with PSYDAC.

yguclu commented:
@FrederikSchnack I have removed the old header lines from all Python files

FrederikSchnack commented on psydac/api/ast/tests/__init__.py:
Should we keep these `__init__.py` files empty?

FrederikSchnack replied:
I'm confused, didn't you agree with this in the comments: "`__init__.py` files should be kept empty"? I think we should just not add unnecessary lines of code (comments).

yguclu replied:
Empty meaning with no Python code in it. Of course we can add docstrings, comments, headers, etc... as we see fit. Sorry for the confusion!

FrederikSchnack commented on psydac/api/tests/__init__.py:
see for other `test/__init__.py`

FrederikSchnack commented on psydac/cmd/__init__.py:
This init file was also empty before. I don't know if we want to keep it that way?

FrederikSchnack commented on psydac/fem/partitioning.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # -*- coding: UTF-8 -*-
```
```suggestion
```

FrederikSchnack commented on psydac/fem/tests/__init__.py:
see above.

yguclu replied:
`__init__.py` files should be kept empty

FrederikSchnack commented on psydac/pyccel/ast/type_inference.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
```suggestion
```

FrederikSchnack commented on psydac/pyccel/codegen/__init__.py:
see above.

FrederikSchnack reviewed (changes requested):
Thank you Yaman for these cumbersome changes! I quickly looked through all of them and commented on a few that were missed. Another question was, if we should put the header on `__init__.py` files that were empty before? I also marked those. (typically with a "see above" :sweat_smile: ) Feel free to resolve the comments yourself as they all are essentially the same.

FrederikSchnack commented on psydac/api/__init__.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # -*- coding: UTF-8 -*-
```
I missed this one in the review.
```suggestion
```

yguclu replied:
d5f3743

yguclu commented on psydac/ddm/tests/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/feec/multipatch/examples/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/mapping/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/polar/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/polar/tests/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/pyccel/symbolic/__init__.py:
`__init__.py` files should be kept empty

yguclu commented on psydac/api/basic.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
d5f3743

yguclu commented on psydac/api/discretization.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
d5f3743

yguclu commented on psydac/core/tests/test_bsplines_kernel.py:
```diff
@@ -1,3 +1,8 @@
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 #coding: utf-8
 
```
cf046d5

yguclu commented on psydac/pyccel/ast/builtins.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
cf046d5

yguclu commented on psydac/pyccel/ast/datatypes.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
cf046d5

yguclu commented on psydac/pyccel/ast/mathext.py:
```diff
@@ -1,3 +1,8 @@
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # Copyright 2020 Yaman Guçlü
 
```
cf046d5

yguclu commented on psydac/pyccel/ast/numpyext.py:
```diff
@@ -1,3 +1,8 @@
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 #!/usr/bin/python
 # -*- coding: utf-8 -*-
```
cf046d5

yguclu commented on psydac/pyccel/codegen/printing/pycode.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding: utf-8
```
cf046d5

FrederikSchnack commented on psydac/cad/__init__.py:
```diff
@@ -1,3 +1,8 @@
+# LICENSE file or go to https://github.com/pyccel/psydac/blob/devel/LICENSE #
+# for full license details.                                                 #
+#---------------------------------------------------------------------------#
 # coding = utf-8
```
This should be the last one (I hope). 
```suggestion
```

FrederikSchnack commented:
For completeness, do you want to apply the script also to the `performance/` and `mesh/` folders? The `performance/test_perf_*.py` also have the `utf-8` header.

yguclu commented:
> For completeness, do you want to apply the script also to the `performance/` and `mesh/` folders? The `performance/test_perf_*.py` also have the `utf-8` header.

- Added new license headers: 95e4ae4
- Removed old header lines: b6159b1

FrederikSchnack reviewed (approved):
Thanks for all these changes, I think this looks good.
